File: backend/app/schemas/experiments/processed.py
"""
Pydantic schemas for Processed entity.

Processed records store calculated/derived results from experiments,
particularly performance metrics like DRE (Decomposition/Removal Efficiency)
and EY (Energy Yield).

Note: The database table has no timestamps - these are calculated results.

Note on imports:
----------------
To avoid circular imports while maintaining proper type serialization,
we use string forward references (e.g., "ExperimentSimple") for nested types.
These are resolved at runtime via model_rebuild() calls.
"""

from __future__ import annotations

from pydantic import BaseModel, Field, ConfigDict
from decimal import Decimal
from typing import Optional, List, TYPE_CHECKING

if TYPE_CHECKING:
    from app.schemas.experiments.experiment import ExperimentSimple


class ProcessedBase(BaseModel):
    """
    Base schema for processed experiment results.
    """

    # Decomposition/Removal Efficiency
    dre: Optional[Decimal] = Field(
        None,
        description="Decomposition/Removal Efficiency (%)",
        examples=["85.5000", "92.3400"]
    )

    # Energy Yield
    ey: Optional[Decimal] = Field(
        None,
        description="Energy Yield (g/kWh or mol/kWh)",
        examples=["12.5000", "8.7500"]
    )


class ProcessedCreate(ProcessedBase):
    """
    Schema for creating a new processed result record.
    """
    pass


class ProcessedUpdate(BaseModel):
    """
    Schema for updating processed results.
    
    All fields optional for partial updates.
    """

    dre: Optional[Decimal] = None
    ey: Optional[Decimal] = None


class ProcessedSimple(BaseModel):
    """
    Simplified schema for nested representations.
    """

    id: int = Field(..., description="Unique identifier")
    dre: Optional[Decimal] = Field(None, description="DRE value")
    ey: Optional[Decimal] = Field(None, description="EY value")

    model_config = ConfigDict(from_attributes=True)


class ProcessedResponse(ProcessedBase):
    """
    Complete schema for processed data returned by the API.
    
    Note: No timestamps since the database table doesn't have them.
    """

    id: int = Field(..., description="Unique identifier")

    # Computed properties
    has_dre: Optional[bool] = Field(
        default=None,
        description="Whether DRE value is recorded"
    )

    has_ey: Optional[bool] = Field(
        default=None,
        description="Whether EY value is recorded"
    )

    is_complete: Optional[bool] = Field(
        default=None,
        description="Whether both metrics are recorded"
    )

    # Optional relationships - using string forward refs
    experiments: Optional[List["ExperimentSimple"]] = Field(
        default=None,
        description="Experiments with these results (included when requested)"
    )

    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "examples": [
                {
                    "id": 1,
                    "dre": "87.5000",
                    "ey": "15.2300",
                    "has_dre": True,
                    "has_ey": True,
                    "is_complete": True
                }
            ]
        }
    )

File: frontend/src/services/api/processed.ts
/**
 * Processed Results API Service
 *
 * Handles CRUD operations for processed/calculated experiment results.
 * Processed results store key performance metrics like DRE (Decomposition/Removal
 * Efficiency) and EY (Energy Yield) that are calculated from raw experiment data.
 *
 * @module services/api/processed
 */

import apiClient from './client';
import type {
    Processed,
    ProcessedCreate,
    ProcessedUpdate,
    ProcessedListParams,
} from './types';

/**
 * Processed Results API methods.
 */
const processedApi = {
    /**
     * List processed results with optional filtering.
     *
     * @param params - Query parameters for filtering
     * @returns Array of processed results
     *
     * @example
     * // Get all complete results
     * const results = await processedApi.list({ complete_only: true });
     *
     * @example
     * // Get results with DRE > 80%
     * const highEfficiency = await processedApi.list({ min_dre: 80 });
     */
    async list(params?: ProcessedListParams): Promise<Processed[]> {
        const response = await apiClient.get<Processed[]>('/api/processed/', { params });
        return response.data;
    },

    /**
     * Get a single processed result by ID.
     *
     * @param id - Processed result ID
     * @param include - Comma-separated relationships to include (e.g., "experiments")
     * @returns The processed result
     *
     * @example
     * const result = await processedApi.get(1, 'experiments');
     */
    async get(id: number, include?: string): Promise<Processed> {
        const response = await apiClient.get<Processed>(`/api/processed/${id}`, {
            params: include ? { include } : undefined,
        });
        return response.data;
    },

    /**
     * Create a new processed result.
     *
     * @param data - Processed result data
     * @returns The created processed result
     *
     * @example
     * const newResult = await processedApi.create({
     *     dre: 85.5,
     *     ey: 12.3,
     * });
     */
    async create(data: ProcessedCreate): Promise<Processed> {
        const response = await apiClient.post<Processed>('/api/processed/', data);
        return response.data;
    },

    /**
     * Update an existing processed result.
     *
     * @param id - Processed result ID
     * @param data - Fields to update
     * @returns The updated processed result
     *
     * @example
     * const updated = await processedApi.update(1, { dre: 90.0 });
     */
    async update(id: number, data: ProcessedUpdate): Promise<Processed> {
        const response = await apiClient.patch<Processed>(`/api/processed/${id}`, data);
        return response.data;
    },

    /**
     * Delete a processed result.
     *
     * @param id - Processed result ID
     *
     * @example
     * await processedApi.delete(1);
     */
    async delete(id: number): Promise<void> {
        await apiClient.delete(`/api/processed/${id}`);
    },
};

export default processedApi;
